File: src/main/java/com/github/kayjamlang/core/provider/MainExpressionProvider.java
package com.github.kayjamlang.core.provider;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.Type;
import com.github.kayjamlang.core.exceptions.TypeException;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

public class MainExpressionProvider<A, B, C> {
    private final Map<Class<? extends Expression>,
            ExpressionProvider<? extends Expression, A, B, C>> providers = new HashMap<>();

    public final A defaultObject;

    public MainExpressionProvider(A defaultObject){
        this.defaultObject = defaultObject;
    }

    public <D extends Expression> void addProvider(
            Class<D> expression,
            ExpressionProvider<D, A, B, C> expressionCompiler) {
        providers.put(expression, expressionCompiler);
    }

    public <D extends Expression> void removeProvider(
            Class<D> expression){
        providers.remove(expression);
    }


    @SuppressWarnings("unchecked")
    public A provide(Expression expression, B context,
                     B argsContext) throws Exception {
        if(providers.containsKey(expression.getClass())){
            ExpressionProvider<? extends Expression, A, B, C> expressionCompiler =
                    providers.get(expression.getClass());
            Method method = expressionCompiler.getClass()
                    .getMethod("provide",
                            MainExpressionProvider.class,
                            Object.class,
                            Object.class,
                            Object.class);

            try {
                return (A)
                        method.invoke(expressionCompiler,
                                this, context, argsContext, expression);
            } catch (InvocationTargetException ite) {
                if(ite.getCause() instanceof Exception)
                    throw (Exception) ite.getCause();
                else ite.getCause().printStackTrace();
            }
        }

        return defaultObject;
    }

    public Type getType(Expression expression, B context,
                                B argsContext) throws TypeException {
        if(providers.containsKey(expression.getClass())){
            ExpressionProvider<? extends Expression, A, B, C> expressionCompiler =
                    providers.get(expression.getClass());
            try {
                Method method = expressionCompiler.getClass()
                        .getMethod("getType",
                                MainExpressionProvider.class,
                                Object.class, Object.class, Object.class);

                return (Type)
                        method.invoke(expressionCompiler,
                                this, context, argsContext, expression);
            } catch (Throwable ignored){}
        }

        return Type.ANY;
    }
}

File: src/main/scala/com/github/kayjamlang/core/provider/MainExpressionProvider.scala
package com.github.kayjamlang.core.provider

import com.github.kayjamlang.core.Type
import com.github.kayjamlang.core.expression.Expression
import com.github.kayjamlang.core.opcodes.AccessType

import java.lang.reflect.InvocationTargetException
import scala.collection.JavaConversions.mapAsJavaMap
import scala.collection.mutable

class MainExpressionProvider[A, B, C](val defaultObject: A, accessType: AccessType) extends Expression(accessType, -1) {
  val providers = new mutable.HashMap[Class[_ <: Expression], ExpressionProvider[_ <: Expression, A, B, C]]

  def addProvider[D <: Expression](expression: Class[D], expressionCompiler: ExpressionProvider[D, A, B, C]): Unit = providers put(expression, expressionCompiler)
  def removeProvider[D <: Expression](expression: Class[D]): Unit = providers remove expression

  @SuppressWarnings(Array("unchecked"))
  @throws[Exception]
  def provide(expression: Expression, context: B, argsContext: B): A = {
    if (providers containsKey expression.getClass) {
      val expressionCompiler = providers get expression.getClass
      val method = expressionCompiler.getClass getMethod("provide", classOf[MainExpressionProvider[_, _, _]], classOf[Any], classOf[Any], classOf[Any])
      try
        return method.invoke(expressionCompiler, this, context.asInstanceOf[Object], argsContext.asInstanceOf[Object], expression).asInstanceOf[A]
      catch {
        case ite: InvocationTargetException =>
          ite getCause match {
            case exception: Exception => throw exception
            case _ => ite.getCause printStackTrace()
          }
      }
    }
    defaultObject
  }

  def getType(expression: Expression, context: B, argsContext: B): Type = {
    if (providers containsKey expression.getClass) {
      val expressionCompiler = providers get expression.getClass
      try {
        val method = expressionCompiler.getClass getMethod("getType", classOf[MainExpressionProvider[_, _, _]], classOf[Any], classOf[Any], classOf[Any])
        return method.invoke(expressionCompiler, this, context.asInstanceOf[Object], argsContext.asInstanceOf[Object], expression).asInstanceOf[Type]
      } catch { case _: Throwable => }
    }
    Type ANY
  }
}
